src/scenic/core/object_types.py

```diff
@@ -3,666 +3,1760 @@
 Defines the 3 Scenic classes `Point`, `OrientedPoint`, and `Object`, and associated
 helper code (notably their base class `Constructible`, which implements the handling of
 property definitions and :ref:`specifier resolution`).
+
+.. warning::
+
+    In 2D compatibility mode, these classes are overwritten with 2D analogs. While
+    we make an effort to map imports to the correct class, this only works if imports
+    use the form ``import scenic.core.object_types as object_types`` followed by accessing
+    ``object_types.Object``. If you instead use ``from scenic.core.object_types import Object``,
+    you may get the wrong class.
+
 """
 
+import typing
+import warnings
 import collections
 import math
 import random
-
-from scenic.core.distributions import Samplable, needsSampling
-from scenic.core.specifiers import Specifier, PropertyDefault
-from scenic.core.vectors import Vector
+import numpy as np
+import trimesh
+from abc import ABC, abstractmethod
+from functools import cache
```

Comment:
`functools.cache` was added in 3.9, so we can't use it. You can use `functools.lru_cache(maxsize=None)` instead. (Although you should think about whether the cache might grow large in practical use, in which case you might want to put a size limit.)